sci-libs/primegen/primegen-0.97.recipe

```diff
@@ -0,0 +1,81 @@
+SUMMARY="Fast C/C++ prime number generator library"
+DESCRIPTION="primegen is a small, fast library to generate prime numbers \
+in order. It generates the 50847534 primes up to 1000000000 in just 8 \
+seconds on a Pentium II-350; it prints them in decimal in just 35 seconds. \
+primegen can generate primes up to 1000000000000000, although it is not \
+optimized for primes past 32 bits. It uses the Sieve of Atkin instead of the \
+traditional Sieve of Eratosthenes."
+HOMEPAGE="https://cr.yp.to/primegen.html"
+COPYRIGHT="1999, D. J. Bernstein"
+LICENSE="Public Domain"
+REVISION="2"
+SOURCE_URI="https://cr.yp.to/primegen/primegen-$portVersion.tar.gz"
+CHECKSUM_SHA256="54285baf8eed9e421ff2220a2112d38cfb20c1ebef6014ef3f0004c22c95f40d"
+SOURCE_DIR="primegen-$portVersion"
+PATCHES="primegen-$portVersion.patchset"
+
+ARCHITECTURES="?x86_gcc2 x86 x86_64"
+
+PROVIDES="
+	primegen = $portVersion
+	cmd:primegaps = $portVersion
+	cmd:primes = $portVersion
+	cmd:primespeed = $portVersion
+	"
+REQUIRES="
+		haiku
+		"
+
+PROVIDES_devel="
+	primegen_devel = $portVersion
+	devel:libprimegen
+	"
+REQUIRES_devel="
+	primegen == $portVersion base
+	"
+
+BUILD_PREREQUIRES="
+		cmd:make
+		cmd:gcc
+		cmd:nroff
+		"
+
+BUILD_REQUIRES="
+		haiku_devel
+		"
+
+BUILD()
+{
+	make
+}
+
+INSTALL()
+{
+	install -d $binDir $docDir $manDir/man1 $manDir/man3
+	install -d $includeDir/primegen $developLibDir $libDir
+	install -t $binDir primegaps primes primespeed
+	install -t $docDir README
+	install -t $manDir/man1 primegaps.1 primes.1 primespeed.1
+	install -t $manDir/man3 primegen.3
+	install -t $includeDir/primegen primegen.h uint32.h uint64.h
+	install -T primegen.a $developLibDir/libprimegen.a
+	install -T primegen.so $developLibDir/libprimegen.so
+	install -T primegen.so $libDir/libprimegen.so
+
+	# devel package
+	packageEntries devel \
+		$developDir \
+		$manDir/man3
+}
+
+TEST()
+{
+	[ "$(primes 1 1000000 | md5sum)" != "c13929ee9d2aea8f83aa076236079e94  -" ] && echo "Test failed!" && exit
+	[ "$(primes 1 10000000 | md5sum)" != "60e34d268bad671a5f299e1ecc988ff6  -" ] && echo "Test failed!" && exit
+	[ "$(primes 1 100000000 | md5sum)" != "4e2b0027288a27e9c99699364877c9db  -" ] && echo "Test failed!" && exit
+	[ "$(primes 1 1000000000 | md5sum)" != "92c178cc5bb85e06366551c0ae7e18f6  -" ] && echo "Test failed!" && exit
+	[ "$(primes 1 10000000000 | md5sum)" != "95ded65c9ddca18e1df966ba2a2b373a  -" ] && echo "Test failed!" && exit
+	[ "$(primes 1 100000000000 | md5sum)" != "9d8a8c2d7221497843ca289e6f463bfd  -" ] && echo "Test failed!" && exit
+	[ "$(primes 1 1000000000000 | md5sum)" != "c899bf5d0ccef05787ac969afd6606ff  -" ] && echo "Test failed!" && exit
+	echo "Test passed!"
```

Comment:
Hmm, I think you can keep only the first test (1000000), or, maybe, the one with 10000000000, if it does not take too long, in order to check that it can handle ints bigger than 2^32.
BTW, I'd use `./primes` instead of just `primes`, just to make sure it will always be found.
You also don't need to have a `echo "Test passed!"` because haikuporter only cares about the return code.
So I 'd use something like:
```
	MD5="`./primes 1 1000000 | md5sum | cut -d" " -f1`"
	test "$MD5" = c13929ee9d2aea8f83aa076236079e94
```
But please feel free to keep your style if you wish. This is only cosmetics ;-)

BTW, if md5sum is not in coreutils, then you need to add `cmd:md5sum` in BUILD_PREREQUIRES.
Hmm, in case you haven't checked whether or not the recipe really needs `cmd:nroff` could you try to build it without nroff?
(It's always better to avoid unneeded dependencies.)

Reply:
I tried it without nroff, the build does in fact need nroff.

Reply:
Does https://www.gnu.org/software/coreutils/manual/html_node/md5sum-invocation.html mean that md5sum is likely in coreutils?

Reply:
Hmm, I guess you had already noticed that `nroff` is needed during the build, to create 4 files:
* primegaps.0
* primegen.0
* primes.0
* primespeed.0

But we don't really need them, so it might be worth patching the `Makefile` to replace
```
it: \
man prog
```
by
```
it: \
prog
```
The quick way to do that is to add a PATCH function before BUILD:
```
PATCH()
{
	# Make "it" depend on "prog" only, instead of "man" and "prog"
	sed -i -e "s/^man prog$/prog/" Makefile
}
```
But it might be better to use the patch file for that.

Let's see what the other contributors think about this.
Maybe, after all, it's better to keep the `cmd:nroff` and ignore the files it generates.

Reply:
The >32-bit test takes a while, so I decided to only keep the smallest one.

When and where is the test run? If it's after the `INSTALL` function, isn't using `primes` to test the correct installation of the `primes` program part of the testing? 😛

Reply:
> Does ... mean that md5sum is likely in coreutils?

Sorry, I should have given the trick:
```
package list -p /system/packages/coreutils-*.hpkg | grep ^bin/md5sum
```
which yelds
```
bin/md5sum
```
So yes, we (also) have it in the coreutils package in Haiku.
This means that we don't need to add a ```cmd:md5sum``` in BUILD_PREREQUIRES because coreutils is always "activated" by haikuporter during the builds.

Reply:
For reference:
```
pkgman search cmd:md5sum
```

Would be more useful if you don't already have an idea where something may come from.

The packagefs also adds virtual attributes to all files that tell what package they are in. You can list them with listattr:

```
listattr -l /bin/md5sum
```

Finally, not adding cmd:md5sum because coreutils happens to be part of the default setup isn't very future proof. A lot of times it's hard to know what implicit tools are used by a certain buildsystem. But if you this obviously need a specific tool, I'd personally add it explicitly.

Reply:
Do the packages in coreutils change often?

Reply:
So I was looking through some other packages and I noticed that they explicitly require `cmd:make`, `cmd:gcc`, and whatnot... is that what you are suggesting I do?

Reply:
I'm quite sure the list of available commands in coreutils never changes.
Regarding the suggested inclusion of `cmd:md5sum` in BUILD_PREREQUIRES, @mmlr was just suggesting that maybe we could decide to change.or remove the (non-written?) rule which allows us to not include such dependencies.
But I guess that removing this rule would cause so much trouble that this will never happen.
As a matter of fact, some of the commands available in coreutils are used in most recipes. So if we wanted to drop this rule we would need to include the `cmd:` lines in really a lot of recipes.
So, don't worry about the primegen's recipe. It's perfect!

Reply:
The fact that "it just works now" is not a reason to ignore this. We may decide to split coreutils in multiple smaller packages, or they may just remove commands from there and we would have to provide replacements in a different package, or we may replace the GPL licenced coreutils by some BSD version, or someone would want to provide a busybox-like all-in-one implementation, or...

And yes, as you said, when this happens it will be a lot of work. But this is not a reason to make it even worse when you know you use a command and don't add a requires for it. By raising awareness about this, we can do the work progressively as we notice such problems, instead of waiting for them to pile up and then having a lot of trouble when they actually start breaking things.

Reply:
Is there a `TEST_PREREQUIRES` that I could use for a command that is only needed for tests?

Reply:
Not at the moment, but we should probably add this to haikuporter.
I suggest the following pattern:

TEST_PREREQUIRES="..."
BUILD_PREREQUIRES += TEST_PREREQUIRES

This way, when HaikuPorter directly supports it, we can simply remove the += line.